File: letter_combinations_of_a_phone_number/LetterCombinationsOfAPhoneNumber.java
class LetterCombinationsOfAPhoneNumber {
    public List<String> letterCombinations(String digits) {
        List<String> letterCombinations = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return letterCombinations;
        }
        
        String[] table = new String[]{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        dfs(digits, table, 0, new StringBuilder(), letterCombinations);
        return letterCombinations;
    }
    
    private void dfs(String digits, String[] table, int lvl, StringBuilder sb, List<String> letterCombinations) {
        if (lvl == digits.length()) {
            letterCombinations.add(sb.toString());
            return;
        }

        int idx = digits.charAt(lvl) - '2';
        String letters = table[idx];
        for (int i = 0; i < letters.length(); i++) {
            char ch = letters.charAt(i);
            sb.append(ch);
            dfs(digits, table, lvl + 1, sb, letterCombinations);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}

File: letter_combinations_of_a_phone_number/Letter_Combinations_Of_A_Phone_Number.cpp
class Letter_Combinations_Of_A_Phone_Number {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) {
            return vector<string>();
        }
        
        vector<vector<char>> map = {{'a', 'b', 'c'}, {'d', 'e', 'f'}, {'g', 'h', 'i'}, {'j', 'k', 'l'}, {'m', 'n', 'o'}, {'p', 'q', 'r', 's'}, {'t', 'u', 'v'}, {'w', 'x', 'y', 'z'}};
        string sol = "";
        vector<string> letter_combinations;
        dfs(digits, 0, sol, map, letter_combinations);        
        return letter_combinations;
    }
    
private:
    void dfs(string digits, int lvl, string& sol, vector<vector<char>>& map, vector<string>& letter_combinations) {
        if (lvl == digits.size()) {
            letter_combinations.push_back(sol);
            return;
        }

        vector<char> chars = map[digits[lvl] - '0' - 2];
        for (char ch : chars) {
            sol.push_back(ch);
            dfs(digits, lvl + 1, sol, map, letter_combinations);
            sol.erase(sol.size() - 1);
        }
    }
};
